CDS-40448-run-stages-in-parallel (#802)
harness/developer-hub: docs/platform/8_Pipelines/run-stages-in-parallel.md

---

> title: Run Stages in Parallel
> description: Run stages in parallel if you want to deploy multiple, independent services to the same environment. 

Reviewer: ```suggestion
title: Run CD stages in parallel
description: Perform parallel stage executions.
```

Reviewer: titles should be in sentence case.
descriptions should be as short as possible.
I removed "deploy" from the description because this applies to any stage type.

> sidebar_position: 3

Reviewer: this setting determines where the topic sits in the left navigation. Be sure to pick a spot that matches the importance of the topic.

---

> Run stages in parallel when you want to deploy multiple services to the same environment. 

Reviewer: ```suggestion
You can run stages in parallel when you want to perform parallel operations. For example:

- CI: Parallel execution of builds and tests can speed up the integration process and provide faster feedback to developers.
- CD: Parallel execution of deployment stages can help reduce deployment times and increase the efficiency of the process.
- Infrastructure as Code (IaC): Parallel execution of IaC scripts can speed up the creation of resources and improve the scalability of infrastructure.
- Testing: Parallel execution of tests can reduce the overall testing time and provide faster feedback to developers.
```

Reviewer: just making it more generic

> ### Requirements

Reviewer: ```suggestion
## Requirements
```

Reviewer: our top-level headings are h2's so ## is what we use.
let me know if that doesn't look right.

* [Learn Harness' Key Concepts](../../getting-started/learn-harness-key-concepts.md)
* [Add a Stage](add-a-stage.md)

> ### Running stages in parallel

Reviewer: ```suggestion
## Running stages in parallel
```

Run stages in parallel by dragging them together or by selecting the `+` button under a stage in the UI. 

![](./static/add-a-stage-57.png)

The only change in the stage YAML is an additional `parallel`. 

```
stages:
    - parallel:
        - stage: 
```

Navigate between the stages by hovering over them and selecting the stage.

![](./static/run-stages-in-parallel.png)

> A [Resource Constraint](../../continuous-delivery/cd-deployments-category/deployment-resource-constraints.md) step is added to every stage automatically to make sure that two infrastructures are not being deployed concurrently. Hence, one of the two parallel stages will deploy first. Once the deployment of the winning stage finishes, the other stages will start.

Reviewer: ```suggestion
A [resource constraint](../../continuous-delivery/cd-deployments-category/deployment-resource-constraints.md) step is added to every stage automatically to make sure that you are not using two infrastructures simultaneously. Hence, one parallel stages executes first. Once the execution of the first stage finishes, the other stage(s) will start.
```

Reviewer: is this true? It's interesting because then it's not really a parallel stage execution. But if the dev signed off, i guess it's true.

Author: This is a suggestion I got from dev when I asked about which stage will execute first.

> You can deploy multiple services and multiple environments in a single stage. However, using multi-service will deploy each selected service to a selected environment. For such use cases, we recommend using multiple stages.

Reviewer: ```suggestion
You can deploy multiple services to multiple environments in a single stage (for more information, go to [Use multiple services and environments in a deployment](https://developer.harness.io/docs/continuous-delivery/cd-deployments-category/multiserv-multienv/)). However, running multi-service stages in parallel will deploy each service to each environment in parallel. This can become very complicated. For such use cases, we recommend using multiple stages.
```

---
title: Run Stages in Parallel
description: Run stages in parallel if you want to deploy multiple, independent services to the same environment. 
sidebar_position: 3
---

> Run stages in parallel when you want to deploy multiple, independent services to the same environment. 

Reviewer 2: even if the services are dependent, they can deploy by not selecting the "deploy services in parallel option". This will deploy the services sequentially
